text/000x-nrd-kernels.md

```diff
@@ -0,0 +1,332 @@
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+An NRD kernel is not valid within a specified number of blocks of a previous duplicate instance of the same NRD kernel. We define duplicate here as two NRD kernels sharing the same public excess commitment. NRD kernels with different excess commitments are not treated as duplicates. An NRD kernel and a non-NRD kernel (plain kernel, coinbase kernel etc.) sharing the same excess commitment are not treated as duplicates.
+
+An NRD kernel has an associated _relative_ lock height. For a block _B_ containing this kernel to be valid, no duplicate instance of the kernel can exist in the last _RH_ blocks (up to and including _B_), where _RH_ is the relative lock height.
+For example, a transaction containing an NRD kernel with relative lock height 1440 (24 hours) is included in a block at height 1000000. This block is only valid if no duplicate instance of this kernel exists in any block from height 998561 (h-1439) to height 1000000 (h-0) inclusive. A duplicate instance _is_ permitted at height 998560 (h-1440), with the transaction seen as valid.
+If no duplicate instance of the kernel exists within this range then the lock criteria is met.
+
+A kernel can be delayed by the existence of a previous kernel. The _non-existence_ of a previous kernel has no impact on the lock criteria. Note that this implies the _first_ singular occurrence of any NRD kernel meets the lock criteria trivially as it cannot, by definition, be locked by a previous kernel. Thus, the relative lock defaults to "fail open" semantics.
+
+Each node maintains an index of _recent_ NRD kernels to enable efficient checking of NRD relative lock heights. Note we only need to index NRD locks and we only need to index those within recent history. Relative locks longer than 7 days are not valid. This is believed to be sufficient to cover all proposed use cases.
+
+The minimum value for a relative lock height is 1 meaning a prior instance of the kernel can exist in the previous block for the lock criteria to be met.
+An instance of the NRD kernel in the _same_ block will invalidate the block as the lock criteria will not be met.
+
+NRD lock heights of 0 are invalid and it is never valid for two duplicate instances of the _same_ NRD kernel to exist in the same block.
+
+It follows that two transactions containing duplicate instances of the same NRD kernel cannot be accepted as valid in the transaction pool concurrently.
+"First one wins" semantics apply when validating transactions containing NRD kernels in a similar way to resolving the spending of unspent outputs.
```

Comment:
A miner is incentivized to resolve txs spending the same outputs by higher fee rather than first seen, and the same should apply to NRD kernels. This allow users to override previous txs by publishing ones with higher fees. (for relay purposes one should require at least a doubling of fees, otherwise too much bandwidth is wasted with repeated replacement with additional nanogrin in fee).

Reply:
Agreed.
We should revisit this when we revisit the fee structure in a separate RFC.
We do not support "replace by fee" today.
The description in this RFC matches our _current_ behavior.

Let me edit this section in the RFC to state this is how things currently work and mention the proposed (as yet unwritten) fee RFC.